In [1]:
import tensorflow as tf
from tensorflow import keras

import numpy as np
import os

np.random.seed(42)
tf.random.set_seed(42)

%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

In [2]:
(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.mnist.load_data()
X_train_full = X_train_full / 255.
X_test = X_test / 255.
X_train, X_valid = X_train_full[:-5000], X_train_full[-5000:]
y_train, y_valid = y_train_full[:-5000], y_train_full[-5000:]

X_train = X_train[..., np.newaxis]
X_valid = X_valid[..., np.newaxis]
X_test = X_test[..., np.newaxis]

#### start: 1:22
#### end: 1:37
#### accuracy: 99.19
#### default notebook

In [ ]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

#### start: 6:19
#### end: 6:26
#### accuracy: 99.04%
#### epochs = 5

In [3]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=5, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/5
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.1890 - accuracy: 0.9415 - val_loss: 0.0517 - val_accuracy: 0.9864
Epoch 2/5
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0819 - accuracy: 0.9750 - val_loss: 0.0402 - val_accuracy: 0.9894
Epoch 3/5
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0601 - accuracy: 0.9813 - val_loss: 0.0330 - val_accuracy: 0.9912
Epoch 4/5
55000/55000 [==============================] - 90s 2ms/sample - loss: 0.0503 - accuracy: 0.9846 - val_loss: 0.0336 - val_accuracy: 0.9910
Epoch 5/5
10000/10000 [==============================] - 3s 274us/sample - loss: 0.0299 - accuracy: 0.9904


[0.029911718372713586, 0.9904]

#### start: 6:28
#### end: 6:51
#### accuracy: 99.23%
#### epochs = 15

In [4]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=15, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/15
55000/55000 [==============================] - 95s 2ms/sample - loss: 0.1890 - accuracy: 0.9415 - val_loss: 0.0517 - val_accuracy: 0.9864
Epoch 2/15
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0819 - accuracy: 0.9750 - val_loss: 0.0402 - val_accuracy: 0.9894
Epoch 3/15
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.0601 - accuracy: 0.9813 - val_loss: 0.0330 - val_accuracy: 0.9912
Epoch 4/15
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0503 - accuracy: 0.9846 - val_loss: 0.0336 - val_accuracy: 0.9910
Epoch 5/15
55000/55000 [==============================] - 92s 2ms/sample - loss: 0.0397 - accuracy: 0.9875 - val_loss: 0.0319 - val_accuracy: 0.9920
Epoch 6/15
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0366 - accuracy: 0.9882 - val_loss: 0.0295 - val_accuracy: 0.9924
Epoch 7/15
55000/55000 [==============================] -

[0.03301791804234945, 0.9923]

#### start: 6:53
#### end: 7:09
#### accuracy: 99.13%

#### keras.layers.Dense(256, activation = "relu")
#### keras.layers.Dropout(0.5)

#### epochs = 10

In [5]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/10
55000/55000 [==============================] - 97s 2ms/sample - loss: 0.2816 - accuracy: 0.9141 - val_loss: 0.0494 - val_accuracy: 0.9852
Epoch 2/10
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.1097 - accuracy: 0.9701 - val_loss: 0.0420 - val_accuracy: 0.9880
Epoch 3/10
55000/55000 [==============================] - 93s 2ms/sample - loss: 0.0891 - accuracy: 0.9746 - val_loss: 0.0373 - val_accuracy: 0.9908
Epoch 4/10
55000/55000 [==============================] - 92s 2ms/sample - loss: 0.0759 - accuracy: 0.9793 - val_loss: 0.0348 - val_accuracy: 0.9900
Epoch 5/10
55000/55000 [==============================] - 91s 2ms/sample - loss: 0.0611 - accuracy: 0.9825 - val_loss: 0.0337 - val_accuracy: 0.9910
Epoch 6/10
55000/55000 [==============================] - 95s 2ms/sample - loss: 0.0527 - accuracy: 0.9850 - val_loss: 0.0345 - val_accuracy: 0.9910
Epoch 7/10
55000/55000 [==============================] -

[0.03138194991455293, 0.9913]

#### start: 7:10
#### end: 7:26
#### accuracy: 99.18%

#### keras.layers.Dense(256, activation = "relu")

#### epochs = 10

In [6]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/10
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.2174 - accuracy: 0.9336 - val_loss: 0.0540 - val_accuracy: 0.9852
Epoch 2/10
55000/55000 [==============================] - 99s 2ms/sample - loss: 0.0897 - accuracy: 0.9739 - val_loss: 0.0413 - val_accuracy: 0.9904
Epoch 3/10
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.0692 - accuracy: 0.9793 - val_loss: 0.0352 - val_accuracy: 0.9912
Epoch 4/10
55000/55000 [==============================] - 94s 2ms/sample - loss: 0.0607 - accuracy: 0.9823 - val_loss: 0.0343 - val_accuracy: 0.9902
Epoch 5/10
55000/55000 [==============================] - 96s 2ms/sample - loss: 0.0505 - accuracy: 0.9848 - val_loss: 0.0288 - val_accuracy: 0.9928
Epoch 6/10
55000/55000 [==============================] - 96s 2ms/sample - loss: 0.0421 - accuracy: 0.9867 - val_loss: 0.0309 - val_accuracy: 0.9916
Epoch 7/10
55000/55000 [==============================] -

[0.026581565553323982, 0.9918]

#### start: 7:29
#### end: 7:50
#### accuracy: 99.12%

#### keras.layers.Dense(256, activation="relu"),
#### keras.layers.Dropout(0.5),
#### keras.layers.Dense(256, activation="relu"),
#### keras.layers.Dropout(0.5),

#### epochs = 10

In [7]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/10
55000/55000 [==============================] - 131s 2ms/sample - loss: 0.2237 - accuracy: 0.9319 - val_loss: 0.0565 - val_accuracy: 0.9842
Epoch 2/10
55000/55000 [==============================] - 127s 2ms/sample - loss: 0.0874 - accuracy: 0.9752 - val_loss: 0.0399 - val_accuracy: 0.9892
Epoch 3/10
55000/55000 [==============================] - 125s 2ms/sample - loss: 0.0640 - accuracy: 0.9816 - val_loss: 0.0310 - val_accuracy: 0.9922
Epoch 4/10
55000/55000 [==============================] - 125s 2ms/sample - loss: 0.0533 - accuracy: 0.9844 - val_loss: 0.0338 - val_accuracy: 0.9918
Epoch 5/10
55000/55000 [==============================] - 125s 2ms/sample - loss: 0.0455 - accuracy: 0.9869 - val_loss: 0.0292 - val_accuracy: 0.9914
Epoch 6/10
55000/55000 [==============================] - 123s 2ms/sample - loss: 0.0392 - accuracy: 0.9887 - val_loss: 0.0347 - val_accuracy: 0.9904
Epoch 7/10
55000/55000 [===========================

[0.03427371580859634, 0.9912]

#### start: 7:50
#### end: 8:38
#### accuracy: 99.18%

#### keras.layers.Conv2D(128, kernel_size=3, padding="same", activation="relu")

#### keras.layers.Dense(256, activation="relu"),
#### keras.layers.Dropout(0.5),
#### keras.layers.Dense(256, activation="relu"),
#### keras.layers.Dropout(0.5),

#### epochs = 10

In [8]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    
    keras.layers.Conv2D(32, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(64, kernel_size=3, padding="same", activation="relu"),
    keras.layers.Conv2D(128, kernel_size=3, padding="same", activation="relu"),
    
    keras.layers.MaxPool2D(),
    keras.layers.Flatten(),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(256, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam",
              metrics=["accuracy"])

model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

model.evaluate(X_test, y_test)

Train on 55000 samples, validate on 5000 samples
Epoch 1/10
55000/55000 [==============================] - 275s 5ms/sample - loss: 0.2011 - accuracy: 0.9389 - val_loss: 0.0525 - val_accuracy: 0.9864
Epoch 2/10
55000/55000 [==============================] - 285s 5ms/sample - loss: 0.0803 - accuracy: 0.9775 - val_loss: 0.0362 - val_accuracy: 0.9904
Epoch 3/10
55000/55000 [==============================] - 296s 5ms/sample - loss: 0.0600 - accuracy: 0.9834 - val_loss: 0.0311 - val_accuracy: 0.9922
Epoch 4/10
55000/55000 [==============================] - 291s 5ms/sample - loss: 0.0480 - accuracy: 0.9860 - val_loss: 0.0303 - val_accuracy: 0.9922
Epoch 5/10
55000/55000 [==============================] - 272s 5ms/sample - loss: 0.0404 - accuracy: 0.9888 - val_loss: 0.0289 - val_accuracy: 0.9928
Epoch 6/10
55000/55000 [==============================] - 298s 5ms/sample - loss: 0.0365 - accuracy: 0.9898 - val_loss: 0.0411 - val_accuracy: 0.9900
Epoch 7/10
55000/55000 [===========================

[0.03396283532655866, 0.9918]